File: csharp/v8-spec/Java/CSharpLexerBase.java
import org.antlr.v4.runtime.*;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashSet;
import java.util.List;
import java.util.Queue;

abstract class CSharpLexerBase extends Lexer
{
    protected CSharpLexerBase(CharStream input)
    {
        super(input);
        initPreprocessor();
    }

    // -------------------------------------------------------------------------
    // Mode-stack helpers
    // -------------------------------------------------------------------------

    public int PeekMode()
    {
        return _modeStack.isEmpty() ? DEFAULT_MODE : _modeStack.peek();
    }

    @Override
    public int popMode()
    {
        if (_modeStack.isEmpty())
        {
            System.err.println("unbalanced ()/{}/[]");
            return DEFAULT_MODE;
        }
        return super.popMode();
    }

    public boolean PeekModeIs(int mode)
    {
        return PeekMode() == mode;
    }

    public boolean LookAheadIs(int pos, int value)
    {
        return getInputStream().LA(pos) == value;
    }

    public boolean LookAheadIsNot(int pos, int value)
    {
        return getInputStream().LA(pos) != value;
    }

    public boolean LookAheadIsRBrace1()    { return getInputStream().LA(1) == '}'; }
    public boolean LookAheadIsNotLBrace2() { return getInputStream().LA(2) != '{'; }
    public boolean PeekModeIsIrsCont()     { return PeekModeIs(CSharpLexer.IRS_CONT); }
    public boolean PeekModeIsIvsCont()     { return PeekModeIs(CSharpLexer.IVS_CONT); }

    public void WrapToken()
    {
        String text = getText();
        setText("\u3014" + text.replace("\u3015", "\u3015\u3015") + "\u3015");
    }

    // -------------------------------------------------------------------------
    // Preprocessor state
    // -------------------------------------------------------------------------
    private final Queue<Token>    _pending   = new ArrayDeque<>();
    private final HashSet<String> _symbols   = new HashSet<>();
    private final Deque<Boolean>  _condition = new ArrayDeque<>();
    private final Deque<Boolean>  _taken     = new ArrayDeque<>();

    // Expression evaluator cursor
    private List<Token> _expr = new ArrayList<>();
    private int         _epos;

    private void initPreprocessor()
    {
        for (String arg : System.getProperty("sun.java.command", "").split(" "))
        {
            if (arg.startsWith("--D"))
            {
                for (String sym : arg.substring(3).split(";"))
                    if (!sym.isEmpty()) _symbols.add(sym);
            }
        }
    }

    private boolean isActive()
    {
        return _condition.isEmpty() || _condition.peek();
    }

    // -------------------------------------------------------------------------
    // nextToken override — intercepts DIRECTIVE-channel tokens
    // -------------------------------------------------------------------------
    @Override
    public Token nextToken()
    {
        if (!_pending.isEmpty()) return _pending.poll();

        Token tok = super.nextToken();

        if (tok.getChannel() == CSharpLexer.DIRECTIVE)
        {
            Token skipped = null;
            int type = tok.getType();
            if      (type == CSharpLexer.DEFINE)   handleDefine();
            else if (type == CSharpLexer.UNDEF)    handleUndef();
            else if (type == CSharpLexer.KW_IF)    skipped = handleIf();
            else if (type == CSharpLexer.ELIF)     skipped = handleElif();
            else if (type == CSharpLexer.KW_ELSE)  skipped = handleElse();
            else if (type == CSharpLexer.ENDIF)    handleEndif();
            if (skipped != null) _pending.add(skipped);
        }

        return tok;
    }

    // -------------------------------------------------------------------------
    // Directive handlers
    // -------------------------------------------------------------------------
    private void handleDefine()
    {
        List<Token> line = collectLine();
        String sym = symbolFromLine(line);
        if (isActive() && sym != null) _symbols.add(sym);
    }

    private void handleUndef()
    {
        List<Token> line = collectLine();
        String sym = symbolFromLine(line);
        if (isActive() && sym != null) _symbols.remove(sym);
    }

    private Token handleIf()
    {
        List<Token> line = collectLine();
        boolean outer  = isActive();
        boolean result = outer && evaluate(line);
        _condition.push(result);
        _taken.push(result);
        return result ? null : skipFalseBlock();
    }

    private Token handleElif()
    {
        List<Token> line = collectLine();
        boolean alreadyTaken = !_taken.isEmpty() ? _taken.pop() : false;
        if (!_condition.isEmpty()) _condition.pop();
        boolean outer  = isActive();
        boolean result = !alreadyTaken && outer && evaluate(line);
        _condition.push(result);
        _taken.push(alreadyTaken || result);
        return result ? null : skipFalseBlock();
    }

    private Token handleElse()
    {
        collectLine();
        boolean alreadyTaken = !_taken.isEmpty() ? _taken.pop() : false;
        if (!_condition.isEmpty()) _condition.pop();
        boolean outer  = isActive();
        boolean result = !alreadyTaken && outer;
        _condition.push(result);
        _taken.push(true);
        return result ? null : skipFalseBlock();
    }

    private void handleEndif()
    {
        collectLine();
        if (!_condition.isEmpty()) _condition.pop();
        if (!_taken.isEmpty())    _taken.pop();
    }

    // -------------------------------------------------------------------------
    // collectLine — drain DIRECTIVE_MODE tokens up to DIRECTIVE_NEW_LINE
    // -------------------------------------------------------------------------
    private List<Token> collectLine()
    {
        List<Token> tokens = new ArrayList<>();
        Token t;
        do
        {
            t = super.nextToken();
            if (t.getChannel() != Lexer.HIDDEN)
                tokens.add(t);
        }
        while (t.getType() != CSharpLexer.DIRECTIVE_NEW_LINE && t.getType() != Token.EOF);
        return tokens;
    }

    private static String symbolFromLine(List<Token> line)
    {
        for (Token t : line)
            if (t.getType() == CSharpLexer.CONDITIONAL_SYMBOL) return t.getText();
        return null;
    }

    // -------------------------------------------------------------------------
    // skipFalseBlock — scan char stream, return SKIPPED_SECTION on HIDDEN channel
    // -------------------------------------------------------------------------
    private Token skipFalseBlock()
    {
        StringBuilder sb   = new StringBuilder();
        CharStream stream   = getInputStream();
        int depth          = 1;
        boolean atLineStart = true;
        int startLine      = getLine();

        while (true)
        {
            int c = stream.LA(1);
            if (c == IntStream.EOF) break;

            if (c == '\r' || c == '\n' || c == 0x85 || c == 0x2028 || c == 0x2029)
            {
                stream.consume();
                sb.append((char)c);
                if (c == '\r' && stream.LA(1) == '\n')
                {
                    stream.consume();
                    sb.append('\n');
                }
                atLineStart = true;
                continue;
            }

            if (atLineStart && (c == ' ' || c == '\t'))
            {
                stream.consume();
                sb.append((char)c);
                continue;
            }

            if (atLineStart && c == '#')
            {
                String kw = peekKeyword(stream);
                if (kw.equals("if"))
                    depth++;
                else if (kw.equals("endif"))
                {
                    if (--depth == 0) break;
                }
                else if ((kw.equals("else") || kw.equals("elif")) && depth == 1)
                    break;
            }

            atLineStart = false;
            stream.consume();
            sb.append((char)c);
        }

        CommonToken tok = new CommonToken(CSharpLexer.SKIPPED_SECTION, sb.toString());
        tok.setChannel(Lexer.HIDDEN);
        tok.setLine(startLine);
        return tok;
    }

    private static String peekKeyword(CharStream stream)
    {
        int i = 2; // LA(1) is '#'
        while (stream.LA(i) == ' ' || stream.LA(i) == '\t') i++;
        StringBuilder sb = new StringBuilder();
        int c;
        while ((c = stream.LA(i)) != -1 && Character.isLetter(c))
        {
            sb.append((char)c);
            i++;
        }
        return sb.toString();
    }

    // -------------------------------------------------------------------------
    // Recursive-descent expression evaluator
    // -------------------------------------------------------------------------
    private boolean evaluate(List<Token> tokens)
    {
        _expr = tokens;
        _epos = 0;
        return parseOr();
    }

    private int peekType()
    {
        if (_epos < _expr.size())
        {
            int t = _expr.get(_epos).getType();
            if (t != CSharpLexer.DIRECTIVE_NEW_LINE && t != Token.EOF) return t;
        }
        return -1;
    }

    private Token eConsume() { return _expr.get(_epos++); }

    private boolean parseOr()
    {
        boolean v = parseAnd();
        while (peekType() == CSharpLexer.TK_OR_OR)  { eConsume(); v = parseAnd() || v; }
        return v;
    }

    private boolean parseAnd()
    {
        boolean v = parseEq();
        while (peekType() == CSharpLexer.TK_AND_AND) { eConsume(); v = parseEq() && v; }
        return v;
    }

    private boolean parseEq()
    {
        boolean v = parseUnary();
        if      (peekType() == CSharpLexer.TK_EQ_EQ) { eConsume(); return v == parseUnary(); }
        else if (peekType() == CSharpLexer.TK_NOT_EQ) { eConsume(); return v != parseUnary(); }
        return v;
    }

    private boolean parseUnary()
    {
        if (peekType() == CSharpLexer.TK_NOT) { eConsume(); return !parseUnary(); }
        return parsePrimary();
    }

    private boolean parsePrimary()
    {
        int t = peekType();
        if (t == CSharpLexer.TRUE)               { eConsume(); return true; }
        if (t == CSharpLexer.FALSE)              { eConsume(); return false; }
        if (t == CSharpLexer.CONDITIONAL_SYMBOL) { return _symbols.contains(eConsume().getText()); }
        if (t == CSharpLexer.TK_LPAREN)
        {
            eConsume();
            boolean v = parseOr();
            if (peekType() == CSharpLexer.TK_RPAREN) eConsume();
            return v;
        }
        return false;
    }
}

File: csharp/v8-spec/CSharp/CSharpLexerBase.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using Antlr4.Runtime;

public abstract class CSharpLexerBase : Lexer
{
    protected CSharpLexerBase(ICharStream input)
        : base(input) { InitPreprocessor(); }

    protected CSharpLexerBase(ICharStream input, TextWriter output, TextWriter errorOutput)
        : base(input, output, errorOutput) { InitPreprocessor(); }

    // ANTLR provides PushMode/PopMode/CurrentMode but no PeekMode.
    // Peek at the top of the mode stack without popping it.
    public int PeekMode()
    {
        return ModeStack.Count == 0 ? DEFAULT_MODE : ModeStack.Peek();
    }

    // Override PopMode to make the stack bottomless: mismatched brackets
    // produce a parser-level error rather than a stack underflow exception.
    public override int PopMode()
    {
        if (ModeStack.Count == 0)
        {
            Console.Error.WriteLine("unbalanced ()/{}/[]");
            return DEFAULT_MODE;
        }
        return base.PopMode();
    }

    // Convenience predicates — names must match Macros.json expansions.
    public bool PeekModeIs(int mode) => PeekMode() == mode;

    public bool LookAheadIs(int pos, int value)    => InputStream.LA(pos) == value;
    public bool LookAheadIsNot(int pos, int value) => InputStream.LA(pos) != value;

    public bool LookAheadIsRBrace1()    => InputStream.LA(1) == '}';
    public bool LookAheadIsNotLBrace2() => InputStream.LA(2) != '{';
    public bool PeekModeIsIrsCont()     => PeekModeIs(CSharpLexer.IRS_CONT);
    public bool PeekModeIsIvsCont()     => PeekModeIs(CSharpLexer.IVS_CONT);

    // Wrap an interpolated-string fragment token in Unicode tortoise brackets
    // U+3014 〔 and U+3015 〕 so that grun tree output remains parseable.
    public void WrapToken()
    {
        Text = "\u3014" + Text.Replace("\u3015", "\u3015\u3015") + "\u3015";
    }

    // -------------------------------------------------------------------------
    // Preprocessor state
    // -------------------------------------------------------------------------
    private readonly Queue<IToken>   _pending   = new Queue<IToken>();
    private readonly HashSet<string> _symbols   = new HashSet<string>(StringComparer.Ordinal);
    private readonly Stack<bool>     _condition = new Stack<bool>(); // is current section active?
    private readonly Stack<bool>     _taken     = new Stack<bool>(); // was any branch taken at this level?

    // Expression evaluator cursor (reused per Evaluate call)
    private IList<IToken> _expr = Array.Empty<IToken>();
    private int           _epos;

    private void InitPreprocessor()
    {
        // Pre-populate from --DSYM or --DSYM;SYM2 command-line arguments (mirrors csc /define:)
        foreach (var arg in Environment.GetCommandLineArgs())
        {
            if (arg.StartsWith("--D", StringComparison.Ordinal))
            {
                foreach (var sym in arg.Substring(3).Split(';'))
                    if (sym.Length > 0) _symbols.Add(sym);
            }
        }
    }

    private bool IsActive() => _condition.Count == 0 || _condition.Peek();

    // -------------------------------------------------------------------------
    // NextToken override — intercepts DIRECTIVE-channel tokens
    // -------------------------------------------------------------------------
    public override IToken NextToken()
    {
        if (_pending.Count > 0) return _pending.Dequeue();

        IToken tok = base.NextToken();

        if (tok.Channel == CSharpLexer.DIRECTIVE)
        {
            IToken skipped = null;
            switch (tok.Type)
            {
                case CSharpLexer.DEFINE:   HandleDefine(); break;
                case CSharpLexer.UNDEF:    HandleUndef();  break;
                case CSharpLexer.KW_IF:    skipped = HandleIf();   break;
                case CSharpLexer.ELIF:     skipped = HandleElif(); break;
                case CSharpLexer.KW_ELSE:  skipped = HandleElse(); break;
                case CSharpLexer.ENDIF:    HandleEndif();  break;
            }
            if (skipped != null) _pending.Enqueue(skipped);
        }

        return tok;
    }

    // -------------------------------------------------------------------------
    // Directive handlers
    // -------------------------------------------------------------------------
    private void HandleDefine()
    {
        var line = CollectLine();
        string sym = SymbolFromLine(line);
        if (IsActive() && sym != null) _symbols.Add(sym);
    }

    private void HandleUndef()
    {
        var line = CollectLine();
        string sym = SymbolFromLine(line);
        if (IsActive() && sym != null) _symbols.Remove(sym);
    }

    private IToken HandleIf()
    {
        var line = CollectLine();
        bool outer = IsActive();
        bool result = outer && Evaluate(line);
        _condition.Push(result);
        _taken.Push(result);
        return result ? null : SkipFalseBlock();
    }

    private IToken HandleElif()
    {
        var line = CollectLine();
        bool alreadyTaken = _taken.Count > 0 ? _taken.Pop() : false;
        if (_condition.Count > 0) _condition.Pop();
        bool outer = IsActive();
        bool result = !alreadyTaken && outer && Evaluate(line);
        _condition.Push(result);
        _taken.Push(alreadyTaken || result);
        return result ? null : SkipFalseBlock();
    }

    private IToken HandleElse()
    {
        CollectLine(); // consume trailing whitespace / newline; no expression
        bool alreadyTaken = _taken.Count > 0 ? _taken.Pop() : false;
        if (_condition.Count > 0) _condition.Pop();
        bool outer = IsActive();
        bool result = !alreadyTaken && outer;
        _condition.Push(result);
        _taken.Push(true); // #else is always the final branch at this level
        return result ? null : SkipFalseBlock();
    }

    private void HandleEndif()
    {
        CollectLine(); // consume trailing whitespace / newline
        if (_condition.Count > 0) _condition.Pop();
        if (_taken.Count > 0)    _taken.Pop();
    }

    // -------------------------------------------------------------------------
    // CollectLine — drain DIRECTIVE_MODE tokens up to and including DIRECTIVE_NEW_LINE
    // -------------------------------------------------------------------------
    private IList<IToken> CollectLine()
    {
        var tokens = new List<IToken>();
        IToken t;
        do
        {
            t = base.NextToken();
            // Drop hidden-channel tokens (DIRECTIVE_WHITESPACES, DIRECTIVE_SINGLE_LINE_COMMENT)
            if (t.Channel != Lexer.Hidden)
                tokens.Add(t);
        }
        while (t.Type != CSharpLexer.DIRECTIVE_NEW_LINE && t.Type != TokenConstants.EOF);
        return tokens;
    }

    private static string SymbolFromLine(IList<IToken> line)
    {
        foreach (var t in line)
            if (t.Type == CSharpLexer.CONDITIONAL_SYMBOL) return t.Text;
        return null;
    }

    // -------------------------------------------------------------------------
    // SkipFalseBlock — scan char stream, return SKIPPED_SECTION on HIDDEN channel
    // -------------------------------------------------------------------------
    private IToken SkipFalseBlock()
    {
        var sb           = new StringBuilder();
        var stream       = (ICharStream)InputStream;
        int depth        = 1;   // nesting depth; we start inside the false #if (depth 1)
        bool atLineStart = true;
        int startLine    = Line;

        while (true)
        {
            int c = stream.LA(1);
            if (c == IntStreamConstants.EOF) break;

            // Newline — reset line-start flag
            if (c == '\r' || c == '\n' || c == 0x85 || c == 0x2028 || c == 0x2029)
            {
                stream.Consume();
                sb.Append((char)c);
                if (c == '\r' && stream.LA(1) == '\n') // \r\n — consume as one newline
                {
                    stream.Consume();
                    sb.Append('\n');
                }
                atLineStart = true;
                continue;
            }

            // Leading whitespace on a fresh line — keep atLineStart true
            if (atLineStart && (c == ' ' || c == '\t'))
            {
                stream.Consume();
                sb.Append((char)c);
                continue;
            }

            // Potential preprocessor directive
            if (atLineStart && c == '#')
            {
                string kw = PeekKeyword(stream);
                if (kw == "if")
                {
                    depth++;
                }
                else if (kw == "endif")
                {
                    if (--depth == 0)
                        break; // leave '#' in stream; HandleEndif pops the stacks
                }
                else if ((kw == "else" || kw == "elif") && depth == 1)
                {
                    break; // leave '#' in stream; HandleElse/HandleElif will handle
                }
                // else: nested #else/#elif inside a deeper #if — consume normally
            }

            atLineStart = false;
            stream.Consume();
            sb.Append((char)c);
        }

        var tok = new CommonToken(CSharpLexer.SKIPPED_SECTION, sb.ToString());
        tok.Channel = Lexer.Hidden;
        tok.Line    = startLine;
        return tok;
    }

    // Peek at the directive keyword after '#' at LA(1) without consuming.
    private static string PeekKeyword(ICharStream stream)
    {
        int i = 2; // LA(1) is '#'
        while (stream.LA(i) == ' ' || stream.LA(i) == '\t') i++;
        var sb = new StringBuilder();
        int c;
        while ((c = stream.LA(i)) != -1 && char.IsLetter((char)c))
        {
            sb.Append((char)c);
            i++;
        }
        return sb.ToString();
    }

    // -------------------------------------------------------------------------
    // Recursive-descent expression evaluator over DIRECTIVE_MODE token list
    // -------------------------------------------------------------------------
    private bool Evaluate(IList<IToken> tokens)
    {
        _expr = tokens;
        _epos = 0;
        return ParseOr();
    }

    private int PeekType()
    {
        if (_epos < _expr.Count)
        {
            int t = _expr[_epos].Type;
            if (t != CSharpLexer.DIRECTIVE_NEW_LINE && t != TokenConstants.EOF) return t;
        }
        return -1;
    }

    private IToken EConsume() => _expr[_epos++];

    private bool ParseOr()
    {
        bool v = ParseAnd();
        while (PeekType() == CSharpLexer.TK_OR_OR)  { EConsume(); v = ParseAnd() || v; }
        return v;
    }

    private bool ParseAnd()
    {
        bool v = ParseEq();
        while (PeekType() == CSharpLexer.TK_AND_AND) { EConsume(); v = ParseEq() && v; }
        return v;
    }

    private bool ParseEq()
    {
        bool v = ParseUnary();
        if      (PeekType() == CSharpLexer.TK_EQ_EQ) { EConsume(); return v == ParseUnary(); }
        else if (PeekType() == CSharpLexer.TK_NOT_EQ) { EConsume(); return v != ParseUnary(); }
        return v;
    }

    private bool ParseUnary()
    {
        if (PeekType() == CSharpLexer.TK_NOT) { EConsume(); return !ParseUnary(); }
        return ParsePrimary();
    }

    private bool ParsePrimary()
    {
        int t = PeekType();
        if (t == CSharpLexer.TRUE)               { EConsume(); return true; }
        if (t == CSharpLexer.FALSE)              { EConsume(); return false; }
        if (t == CSharpLexer.CONDITIONAL_SYMBOL) { return _symbols.Contains(EConsume().Text); }
        if (t == CSharpLexer.TK_LPAREN)
        {
            EConsume();
            bool v = ParseOr();
            if (PeekType() == CSharpLexer.TK_RPAREN) EConsume();
            return v;
        }
        return false; // malformed expression
    }
}

